# Analyzing NYC SAT scores

In [3]:
import pandas as pd
import numpy as np

In [6]:
# Reading in all the data in dictionary
data_files = [
    "ap_2010.csv",
    "class_size.csv",
    "demographics.csv",
    "graduation.csv",
    "hs_directory.csv",
    "sat_results.csv"
]

data = {}
    
for f in data_files:
    d = pd.read_csv("schools/{0}".format(f))
    key_name = f.replace(".csv", "")
    data[key_name] = d

** Let's explore SAT results as it contains information for every high school in New York City **

In [8]:
data["sat_results"].head(5)

DBN                                    SCHOOL NAME  \
0  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL STUDIES   
1  01M448            UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   
2  01M450                     EAST SIDE COMMUNITY SCHOOL   
3  01M458                      FORSYTH SATELLITE ACADEMY   
4  01M509                        MARTA VALLE HIGH SCHOOL   

  Num of SAT Test Takers SAT Critical Reading Avg. Score SAT Math Avg. Score  \
0                     29                             355                 404   
1                     91                             383                 423   
2                     70                             377                 402   
3                      7                             414                 401   
4                     44                             390                 433   

  SAT Writing Avg. Score  
0                    363  
1                    366  
2                    370  
3                    359  
4                    384

From SAT results data, it seems like `DBN` is unique for each high school. We have scores for for Reading, Writing and Math.

In [10]:
for k,v in data.items():
    print("\n" + k + "\n")
    print(v.head(3))


graduation

    Demographic     DBN                            School Name Cohort  \
0  Total Cohort  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL   2003   
1  Total Cohort  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL   2004   
2  Total Cohort  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL   2005   

   Total Cohort Total Grads - n Total Grads - % of cohort Total Regents - n  \
0             5               s                         s                 s   
1            55              37                     67.3%                17   
2            64              43                     67.2%                27   

  Total Regents - % of cohort Total Regents - % of grads  \
0                           s                          s   
1                       30.9%                      45.9%   
2                       42.2%                      62.8%   

             ...            Regents w/o Advanced - n  \
0            ...                                   s   
1            ...        

`DBN` is available as column with unique values in every dataframe. This is how, we'll combine all of our datasets into one keeping `DBN` as primary key value. 

** Reading survey data **

Survey data is available in tab delimited formate. Hence, we'll be using encoding value to read it. Once done, let's combine to single dataframe to make it easier to read.

In [12]:
all_survey = pd.read_csv("schools/survey_all.txt", delimiter="\t", encoding="windows-1252")
d75_survey = pd.read_csv("schools/survey_d75.txt", delimiter="\t", encoding="windows-1252")

survey = pd.concat([all_survey, d75_survey], axis=0)
survey.head(3)

N_p  N_s   N_t  aca_p_11  aca_s_11  aca_t_11  aca_tot_11    bn  com_p_11  \
0   90.0  NaN  22.0       7.8       NaN       7.9         7.9  M015       7.6   
1  161.0  NaN  34.0       7.8       NaN       9.1         8.4  M019       7.6   
2  367.0  NaN  42.0       8.6       NaN       7.5         8.0  M020       8.3   

   com_s_11   ...    t_q8c_1  t_q8c_2  t_q8c_3 t_q8c_4  t_q9  t_q9_1  t_q9_2  \
0       NaN   ...       29.0     67.0      5.0     0.0   NaN     5.0    14.0   
1       NaN   ...       74.0     21.0      6.0     0.0   NaN     3.0     6.0   
2       NaN   ...       33.0     35.0     20.0    13.0   NaN     3.0     5.0   

   t_q9_3  t_q9_4  t_q9_5  
0    52.0    24.0     5.0  
1     3.0    78.0     9.0  
2    16.0    70.0     5.0  

[3 rows x 2773 columns]

Since there are 2773 columns, let's get only necessary columns based off [data dictionary from download location](https://data.cityofnewyork.us/Education/NYC-School-Survey-2011/mnz3-dyi8)

In [13]:
# Changing case of dbn to make it easier to merge
survey["DBN"] = survey["dbn"]

# Required columns
columns = ["DBN", "rr_s", "rr_t", "rr_p", "N_s", "N_t", "N_p", "saf_p_11", "com_p_11", "eng_p_11", "aca_p_11", "saf_t_11", "com_t_11", "eng_t_10", "aca_t_11", "saf_s_11", "com_s_11", "eng_s_11", "aca_s_11", "saf_tot_11", "com_tot_11", "eng_tot_11", "aca_tot_11",]

survey= survey.loc[:,columns]
data["survey"] = survey

survey.head(3)

DBN  rr_s  rr_t  rr_p  N_s   N_t    N_p  saf_p_11  com_p_11  eng_p_11  \
0  01M015   NaN    88    60  NaN  22.0   90.0       8.5       7.6       7.5   
1  01M019   NaN   100    60  NaN  34.0  161.0       8.4       7.6       7.6   
2  01M020   NaN    88    73  NaN  42.0  367.0       8.9       8.3       8.3   

      ...      eng_t_10  aca_t_11  saf_s_11  com_s_11  eng_s_11  aca_s_11  \
0     ...           NaN       7.9       NaN       NaN       NaN       NaN   
1     ...           NaN       9.1       NaN       NaN       NaN       NaN   
2     ...           NaN       7.5       NaN       NaN       NaN       NaN   

   saf_tot_11  com_tot_11  eng_tot_11  aca_tot_11  
0         8.0         7.7         7.5         7.9  
1         8.5         8.1         8.2         8.4  
2         8.2         7.3         7.5         8.0  

[3 rows x 23 columns]

** Cleaning up other data files **

Some of the other data files need to cleaned up as they don't have `DBN` column.

In [14]:
data["class_size"].head(3)

CSD BOROUGH SCHOOL CODE                SCHOOL NAME GRADE  PROGRAM TYPE  \
0    1       M        M015  P.S. 015 Roberto Clemente     0K       GEN ED   
1    1       M        M015  P.S. 015 Roberto Clemente     0K          CTT   
2    1       M        M015  P.S. 015 Roberto Clemente     01       GEN ED   

  CORE SUBJECT (MS CORE and 9-12 ONLY) CORE COURSE (MS CORE and 9-12 ONLY)  \
0                                    -                                   -   
1                                    -                                   -   
2                                    -                                   -   

  SERVICE CATEGORY(K-9* ONLY)  NUMBER OF STUDENTS / SEATS FILLED  \
0                           -                               19.0   
1                           -                               21.0   
2                           -                               17.0   

   NUMBER OF SECTIONS  AVERAGE CLASS SIZE  SIZE OF SMALLEST CLASS  \
0                 1.0                19.0                    19.0   
1                 1.0                21.0                    21.0   
2                 1.0                17.0                    17.0   

   SIZE OF LARGEST CLASS DATA SOURCE  SCHOOLWIDE PUPIL-TEACHER RATIO  
0                   19.0         ATS                             NaN  
1                   21.0         ATS                             NaN  
2                   17.0         ATS                             NaN

It can be seen that `DBN` column can be created by combining `CSD` and `SCHOOL CODE` columns.

In [16]:
data["hs_directory"]["DBN"] = data["hs_directory"]["dbn"]

def padded_csd(num):
    string_num = str(num)
    if len(string_num) > 1:
        return string_num
    else:
        return "0" + string_num
    
data["class_size"]["padded_csd"] = data["class_size"]["CSD"].apply(padded_csd)
data["class_size"]["DBN"] = data["class_size"]["padded_csd"] + data["class_size"]["SCHOOL CODE"]

data["class_size"].head(3)

CSD BOROUGH SCHOOL CODE                SCHOOL NAME GRADE  PROGRAM TYPE  \
0    1       M        M015  P.S. 015 Roberto Clemente     0K       GEN ED   
1    1       M        M015  P.S. 015 Roberto Clemente     0K          CTT   
2    1       M        M015  P.S. 015 Roberto Clemente     01       GEN ED   

  CORE SUBJECT (MS CORE and 9-12 ONLY) CORE COURSE (MS CORE and 9-12 ONLY)  \
0                                    -                                   -   
1                                    -                                   -   
2                                    -                                   -   

  SERVICE CATEGORY(K-9* ONLY)  NUMBER OF STUDENTS / SEATS FILLED  \
0                           -                               19.0   
1                           -                               21.0   
2                           -                               17.0   

   NUMBER OF SECTIONS  AVERAGE CLASS SIZE  SIZE OF SMALLEST CLASS  \
0                 1.0                19.0                    19.0   
1                 1.0                21.0                    21.0   
2                 1.0                17.0                    17.0   

   SIZE OF LARGEST CLASS DATA SOURCE  SCHOOLWIDE PUPIL-TEACHER RATIO  \
0                   19.0         ATS                             NaN   
1                   21.0         ATS                             NaN   
2                   17.0         ATS                             NaN   

  padded_csd     DBN  
0         01  01M015  
1         01  01M015  
2         01  01M015

** Combining all the 3 SAT scores to one **

In [21]:
# Using pandas.to_numeric to convert values to numeric 
courses = ["SAT Math Avg. Score", "SAT Critical Reading Avg. Score", "SAT Writing Avg. Score"]

for c in courses:
    data["sat_results"][c] = pd.to_numeric(data["sat_results"][c], errors="coerce")

data["sat_results"]["sat_score"] = data["sat_results"][courses[0]] + data["sat_results"][courses[1]] + data["sat_results"][courses[2]]

data["sat_results"]["sat_score"].head(3)

0    1122.0
1    1172.0
2    1149.0
Name: sat_score, dtype: float64

** Getting coordinates for each school for maps **

In [23]:
import re

def get_latitude(address):
    lat_long =  re.findall("\(.+, .+\)", address)
    return lat_long[0].split(", ")[0].replace("(", "")
    
data["hs_directory"]["lat"] = data["hs_directory"]["Location 1"].apply(get_latitude)

def get_longitude(address):
    lat_long =  re.findall("\(.+, .+\)", address)
    return lat_long[0].split(",")[1].replace(")", "").strip()
    
data["hs_directory"]["lon"] = data["hs_directory"]["Location 1"].apply(get_longitude)

data["hs_directory"]["lat"] = pd.to_numeric(data["hs_directory"]["lat"], errors="coerce")
data["hs_directory"]["lon"] = pd.to_numeric(data["hs_directory"]["lon"], errors="coerce")

data["hs_directory"].head(3)

dbn                                  school_name      boro  \
0  17K548          Brooklyn School for Music & Theatre  Brooklyn   
1  09X543             High School for Violin and Dance     Bronx   
2  09X327  Comprehensive Model School Project M.S. 327     Bronx   

  building_code  phone_number    fax_number grade_span_min  grade_span_max  \
0          K440  718-230-6250  718-230-6262              9              12   
1          X400  718-842-0687  718-589-9849              9              12   
2          X240  718-294-8111  718-294-8109              6              12   

  expgrade_span_min  expgrade_span_max    ...      \
0               NaN                NaN    ...       
1               NaN                NaN    ...       
2               NaN                NaN    ...       

                        priority05 priority06 priority07 priority08  \
0                              NaN        NaN        NaN        NaN   
1                              NaN        NaN        NaN        NaN   
2  Then to New York City residents        NaN        NaN        NaN   

  priority09  priority10                                         Location 1  \
0        NaN         NaN  883 Classon Avenue\nBrooklyn, NY 11225\n(40.67...   
1        NaN         NaN  1110 Boston Road\nBronx, NY 10456\n(40.8276026...   
2        NaN         NaN  1501 Jerome Avenue\nBronx, NY 10452\n(40.84241...   

      DBN        lat        lon  
0  17K548  40.670299 -73.961648  
1  09X543  40.827603 -73.904475  
2  09X327  40.842414 -73.916162  

[3 rows x 61 columns]